play-json/shared/src/main/scala/JsMacroImpl.scala

```diff
@@ -667,33 +674,49 @@ import scala.reflect.macros.blackbox
 
       val syntaxImport = if (!multiParam && !writes) q"" else q"import $syntax._"
       @inline def buildCall = q"$canBuild.$applyOrMap(..${conditionalList(applyFunction, utility.unapplyFunction)})"
-      def readResult =
-        if (multiParam) q"underlying.reads(obj)"
-        else q"underlying.flatMap[${atag.tpe}]($json.Reads.pure(_)).reads(obj)"
+      def finalReads(underlying: TermName) = {
+        val theReads = {
+          if (multiParam) q"$underlying"
+          else q"$underlying.flatMap[${atag.tpe}]($json.Reads.pure(_))"
+        }
+        if (hasOption[StrictProperties]) {
+          val properties = resolvedImplicits.map(ri => configNaming(ri._1))
+          // Directly instantiate FunctionalBuilderOps because we can't rely on function syntax being imported
+          q"""{
+            new $libs.functional.FunctionalBuilderOps($json.Reads.failOnUnknownProperties(
+              _root_.scala.collection.immutable.Seq(..$properties)
+            )).and($theReads: $json.Reads[${atag.tpe}]).apply((_, result) => result)
+          }"""
+        } else {
+          theReads
+        }
+      }
 
       val canBuildCall = methodName match {
-        case "read" => {
+        case "read" =>
+          val underlying = TermName(c.freshName("underlying"))
+          val obj = TermName(c.freshName("obj"))
```

Comment:
Just a small hygiene change, macros shouldn't just define arbitrary terms, they should synthesis them with freshName.